dynamodb/batch/batch.go

```diff
@@ -0,0 +1,150 @@
+	SetTableName(tableName string)
+	SetRequestLimit(newLimit int) error
+	SetMaxTimeout(newTimeout int64) error
+	Add(requestType RequestType, item map[string]*dynamodb.AttributeValue) ([]*dynamodb.ConsumedCapacity, error)
+	Send() ([]*dynamodb.ConsumedCapacity, error)
+}
+
+// NewBatch returns a new object with the given table name, request limit, and dynamo API
+func NewBatch(ddb dynamodbiface.DynamoDBAPI) *Batch {
+	return &Batch{
+		dynamoapi:    ddb,
+		sleepSeconds: 1,
+		requestLimit: requestLimit,
+		maxTimeout:   maxTimeout,
+		bwInput:      basicBatchInput(),
+	}
+}
+
+func basicBatchInput() dynamodb.BatchWriteItemInput {
+	return dynamodb.BatchWriteItemInput{
+		RequestItems:           map[string][]*dynamodb.WriteRequest{},
+		ReturnConsumedCapacity: aws.String("TOTAL"),
+	}
+}
+
+// SetTableName sets the table name
+func (b *Batch) SetTableName(tableName string) {
+	b.table = tableName
+	return
+}
+
+// SetRequestLimit sets the number of items to pile up before sending
+func (b *Batch) SetRequestLimit(newLimit int) error {
+	if newLimit > requestLimit {
+		return fmt.Errorf("requestLimit must be %d or less", requestLimit)
+	}
+	if newLimit < 1 {
+		return fmt.Errorf("requestLimit must be %d or less", requestLimit)
+	}
+	b.requestLimit = newLimit
+	return nil
+}
+
+// SetMaxTimeout sets the maximum sleep allowed for retries
+func (b *Batch) SetMaxTimeout(newTimeout int64) error {
+	if newTimeout > maxTimeout {
+		return fmt.Errorf("too much timeout, should be less than %d", maxTimeout)
+	}
+	if newTimeout < 1 {
+		return fmt.Errorf("not enough time out, should be at least 1")
+	}
+	b.maxTimeout = int64(newTimeout)
+	return nil
+}
+
+// Add an item to the batch, and send the batch if it's full
+func (b *Batch) Add(requestType RequestType, item map[string]*dynamodb.AttributeValue) ([]*dynamodb.ConsumedCapacity, error) {
+	if b.table == "" {
+		return nil, fmt.Errorf("table name required, call SetTableName")
+	}
+	request := &dynamodb.WriteRequest{}
+	switch requestType {
+	case PutRequest:
+		request.PutRequest = &dynamodb.PutRequest{Item: item}
+	case DeleteRequest:
+		request.DeleteRequest = &dynamodb.DeleteRequest{Key: item}
+	}
+	b.requests = append(b.requests, request)
+	if len(b.requests) >= b.requestLimit {
+		return b.Send()
+	}
+	return nil, nil
+}
+
+// Send the batch, if there's anything in it
+func (b *Batch) Send() ([]*dynamodb.ConsumedCapacity, error) {
+	if len(b.requests) < 1 {
+		return nil, nil
+	}
+	if b.sleepSeconds > b.maxTimeout {
+		return nil, fmt.Errorf("we timed out - maxTimeout is %d", b.maxTimeout)
+	}
+	b.bwInput.RequestItems[b.table] = b.requests
+	out, err := b.dynamoapi.BatchWriteItem(&b.bwInput)
+	if err != nil {
+		return nil, err
+	}
+	b.requests = out.UnprocessedItems[b.table]
+	if len(b.requests) > 0 {
+		b.sleepSeconds++
+		time.Sleep(time.Duration(b.sleepSeconds))
+		return b.Send()
+	}
+	b.sleepSeconds = 0
+	delete(b.bwInput.RequestItems, b.table)
+	return out.ConsumedCapacity, nil
```

Comment:
Won't this just return the latest consumed capacity rather than the sum consumed by all the recursive calls to `BatchWriteItem`?

Reply:
Yep, probably. This was a nice-to-have, though. I guess I'll accumulate it instead and return the total.